common/archiver/s3store/historyArchiver.go

```diff
@@ -168,26 +178,61 @@ func (h *historyArchiver) Archive(
 			return archiver.ErrHistoryMutated
 		}
 
-		historyBatches = append(historyBatches, historyBlob.Body...)
-	}
+		encodedHistoryPart, err := encode(historyBlob.Body)
+		if err != nil {
+			logger.Error(archiver.ArchiveNonRetriableErrorMsg, tag.ArchivalArchiveFailReason(errEncodeHistory), tag.Error(err))
+			return err
+		}
 
-	encodedHistoryBatches, err := encode(historyBatches)
-	if err != nil {
-		logger.Error(archiver.ArchiveNonRetriableErrorMsg, tag.ArchivalArchiveFailReason(errEncodeHistory), tag.Error(err))
-		return err
+		key := constructHistoryKey(URI.Path(), request.DomainID, request.WorkflowID, request.RunID, request.CloseFailoverVersion, progress.BatchIdx)
+		if err := upload(ctx, h.s3cli, URI, key, encodedHistoryPart); err != nil {
+			logger := logger.WithTags(tag.ArchivalArchiveFailReason(errWriteKey), tag.Error(err))
+			if isRetryableError(err) {
+				logger.Error(archiver.ArchiveTransientErrorMsg)
+			} else {
+				logger.Error(archiver.ArchiveNonRetriableErrorMsg)
+			}
+			return err
+		}
+		uploadedSize := int64(binary.Size(encodedHistoryPart))
+		scope.AddCounter(metrics.HistoryArchiverTotalUploadSize, uploadedSize)
+		scope.AddCounter(metrics.HistoryArchiverHistorySize, uploadedSize)
```

Comment:
It's actually not easy to emit metrics for historySize and total upload size, consider the following two situations:

1. After having uploaded two blobs, the worker crash. Later when worker is restarted, it will start from the third blob, but the history size counter will start from 0. To solve this, the size should also be added to the progress.
2. If you added existence check before uploading, then the size of skipped blob counts to the history size but not the upload size.

Feel free to just leave TODO here without emitting those two metrics if it's too much work.

Reply:
Nice, am I doing it right now?

Reply:
👍 Yes, it's correct!